internal/pipeline/executor.go

```diff
@@ -281,3 +313,46 @@ func fnChain(pl *kptfilev1alpha2.Pipeline, pkgPath pkg.UniquePath) ([]kio.Filter
 	}
 	return runners, nil
 }
+
+// trackInputFiles records file paths of input resources in the hydration context.
+func trackInputFiles(hctx *hydrationContext, input []*yaml.RNode) error {
+	if hctx.inputFiles == nil {
+		hctx.inputFiles = sets.String{}
+	}
+	for _, r := range input {
+		path, _, err := kioutil.GetFileAnnotations(r)
+		if err != nil {
+			return fmt.Errorf("path annotation missing: %w", err)
+		}
+		hctx.inputFiles.Insert(path)
+	}
+	return nil
+}
+
+// trackOutputfiles records the file paths of output resources in the hydration
+// context. It should be invoked post hydration.
+func trackOutputFiles(hctx *hydrationContext) error {
+	outputSet := sets.String{}
+
+	for _, r := range hctx.root.resources {
+		path, _, err := kioutil.GetFileAnnotations(r)
+		if err != nil {
+			return fmt.Errorf("path annotation missing: %w", err)
+		}
+		outputSet.Insert(path)
+	}
+	hctx.outputFiles = outputSet
+	return nil
+}
+
+// pruneResources compares the input and output of the hydration and prunes
+// resources that are no longer present in the output of the hydration.
+func pruneResources(hctx *hydrationContext) error {
+	resourcesTobePruned := hctx.inputFiles.Difference(hctx.outputFiles)
+	for f := range resourcesTobePruned {
+		if err := os.Remove(filepath.Join(string(hctx.root.pkg.UniquePath), f)); err != nil {
+			return fmt.Errorf("failed to delete file: %w", err)
```

Comment:
Pruning is not file-based by resource-based. In a multi-object files, a file is only deleted if all the resources in that file are deleted.